File: src/screensaver/capture.py
"""Screen capture utilities using mss and Pillow."""

from __future__ import annotations

from typing import Optional, Tuple

import mss
import mss.tools
from PIL import Image


# A region is expressed as (left, top, width, height) in screen pixels.
Region = Tuple[int, int, int, int]


def capture_full_screen(monitor: int = 1) -> Image.Image:
    """Capture the full contents of *monitor* (1-based index) and return a
    Pillow :class:`~PIL.Image.Image`.

    Args:
        monitor: 1-based monitor index.  ``1`` is the primary monitor.

    Returns:
        A ``PIL.Image`` in ``RGB`` mode.
    """
    with mss.mss() as sct:
        raw = sct.grab(sct.monitors[monitor])
    return Image.frombytes("RGB", raw.size, raw.bgra, "raw", "BGRX")


def capture_region(region: Region, monitor: int = 1) -> Image.Image:
    """Capture a rectangular *region* of *monitor* and return a Pillow image.

    Args:
        region: ``(left, top, width, height)`` in screen pixels, relative to
                the top-left corner of *monitor*.
        monitor: 1-based monitor index.

    Returns:
        A ``PIL.Image`` in ``RGB`` mode.

    Raises:
        ValueError: If *width* or *height* is not positive.
    """
    left, top, width, height = region
    if width <= 0 or height <= 0:
        raise ValueError(
            f"Region width and height must be positive, got ({width}, {height})"
        )

    with mss.mss() as sct:
        mon = sct.monitors[monitor]
        bbox = {
            "left": mon["left"] + left,
            "top": mon["top"] + top,
            "width": width,
            "height": height,
        }
        raw = sct.grab(bbox)
    return Image.frombytes("RGB", raw.size, raw.bgra, "raw", "BGRX")


def get_monitor_info(monitor: int = 1) -> dict:
    """Return geometry information for *monitor*.

    Returns a dict with keys ``left``, ``top``, ``width``, ``height``.
    """
    with mss.mss() as sct:
        return dict(sct.monitors[monitor])


def get_monitor_count() -> int:
    """Return the number of available monitors (excluding the virtual 'all' entry)."""
    with mss.mss() as sct:
        # monitors[0] is the combined virtual screen; real monitors start at index 1
        return max(0, len(sct.monitors) - 1)


def normalize_region(
    x1: int, y1: int, x2: int, y2: int
) -> Optional[Region]:
    """Convert two arbitrary corner points into a normalised ``(left, top, width, height)``
    region tuple, ensuring *width* and *height* are positive.

    Returns ``None`` if the two points are identical (zero-area selection).
    """
    left = min(x1, x2)
    top = min(y1, y2)
    width = abs(x2 - x1)
    height = abs(y2 - y1)
    if width == 0 or height == 0:
        return None
    return (left, top, width, height)

File: internal/capture/capture.go
// Package capture provides screen capture utilities.
//
// It wraps the kbinani/screenshot library to capture full-screen or
// rectangular regions across multiple monitors.
package capture

import (
	"fmt"
	"image"

	"github.com/kbinani/screenshot"
)

// Region represents a rectangular area on screen as (X, Y, Width, Height).
type Region struct {
	X, Y, Width, Height int
}

// FullScreen captures the full contents of the given monitor (0-based index)
// and returns an *image.RGBA.
//
// Pass -1 to capture the combined virtual screen (all monitors).
func FullScreen(monitor int) (*image.RGBA, error) {
	n := screenshot.NumActiveDisplays()
	if n == 0 {
		return nil, fmt.Errorf("no active displays found")
	}

	if monitor == -1 {
		// Capture all monitors combined.
		bounds := image.Rect(0, 0, 0, 0)
		for i := 0; i < n; i++ {
			b := screenshot.GetDisplayBounds(i)
			bounds = bounds.Union(b)
		}
		return screenshot.Capture(bounds.Min.X, bounds.Min.Y, bounds.Dx(), bounds.Dy())
	}

	if monitor < 0 || monitor >= n {
		return nil, fmt.Errorf("monitor index %d out of range (0..%d)", monitor, n-1)
	}

	bounds := screenshot.GetDisplayBounds(monitor)
	return screenshot.Capture(bounds.Min.X, bounds.Min.Y, bounds.Dx(), bounds.Dy())
}

// CaptureRegion captures a rectangular region from the given monitor.
//
// The region coordinates are relative to the top-left corner of the monitor.
func CaptureRegion(region Region, monitor int) (*image.RGBA, error) {
	if region.Width <= 0 || region.Height <= 0 {
		return nil, fmt.Errorf("region width and height must be positive, got (%d, %d)", region.Width, region.Height)
	}

	n := screenshot.NumActiveDisplays()
	if n == 0 {
		return nil, fmt.Errorf("no active displays found")
	}
	if monitor < 0 || monitor >= n {
		return nil, fmt.Errorf("monitor index %d out of range (0..%d)", monitor, n-1)
	}

	bounds := screenshot.GetDisplayBounds(monitor)
	return screenshot.Capture(
		bounds.Min.X+region.X,
		bounds.Min.Y+region.Y,
		region.Width,
		region.Height,
	)
}

// MonitorInfo returns geometry information for a monitor (0-based index).
func MonitorInfo(monitor int) (image.Rectangle, error) {
	n := screenshot.NumActiveDisplays()
	if n == 0 {
		return image.Rectangle{}, fmt.Errorf("no active displays found")
	}
	if monitor < 0 || monitor >= n {
		return image.Rectangle{}, fmt.Errorf("monitor index %d out of range (0..%d)", monitor, n-1)
	}
	return screenshot.GetDisplayBounds(monitor), nil
}

// MonitorCount returns the number of active monitors.
func MonitorCount() int {
	return screenshot.NumActiveDisplays()
}

// NormalizeRegion converts two arbitrary corner points into a normalised
// Region with positive Width and Height. Returns nil if the selection
// has zero area.
func NormalizeRegion(x1, y1, x2, y2 int) *Region {
	left := min(x1, x2)
	top := min(y1, y2)
	width := abs(x2 - x1)
	height := abs(y2 - y1)

	if width == 0 || height == 0 {
		return nil
	}

	return &Region{X: left, Y: top, Width: width, Height: height}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
